src/Algebra/Graph/AdjacencyIntMap/Algorithm.hs

```diff
@@ -176,10 +188,26 @@ dfsForest = Typed.dfsForest . Typed.fromAdjacencyIntMap
 -- 'forest' (dfsForestFrom [3] $ 'circuit' [1..5] + 'circuit' [5,4..1]) == 'path' [3,2,1,5,4]
 -- @
 dfsForestFrom :: [Int] -> AdjacencyIntMap -> Forest Int
-dfsForestFrom vs = Typed.dfsForestFrom vs . Typed.fromAdjacencyIntMap
+dfsForestFrom vs g = dfsForestFrom' [ v | v <- vs, hasVertex v g ] g
+
+dfsForestFrom' :: [Int] -> AdjacencyIntMap -> Forest Int
+dfsForestFrom' vs g = evalState (explore vs) IntSet.empty where
+  explore (v:vs) = discovered v >>= \case
+    True -> (:) <$> walk v <*> explore vs
+    False -> explore vs
+  explore [] = return []
+  walk v = Node v <$> explore (adjacent v)
+  adjacent v = IntSet.toList (postIntSet v g)
+  discovered v = do new <- gets (not . IntSet.member v)
+                    when new $ modify' (IntSet.insert v)
```

Comment:
Just curiosity: if you used a non-strict `modify` here, would that lead to poor performance?

Reply:
I tried it out, and the difference seems to be pretty negligible:

[https://github.com/jitwit/alga/blob/bench-dfs/bench/current_results.txt](https://github.com/jitwit/alga/blob/bench-dfs/bench/current_results.txt)

Might lazy be the better choice (even if there's a small hit)?

Reply:
Let's keep it as is. Thanks for checking!